codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -285,193 +321,509 @@ private void writeFromStreamChannel0(Object msg, boolean flush, ChannelPromise p
+        private final class Http2ChannelUnsafe extends AbstractUnsafe {
+            @Override
+            public void connect(final SocketAddress remoteAddress,
+                                SocketAddress localAddress, final ChannelPromise promise) {
+                promise.setFailure(new UnsupportedOperationException());
+            }
         }
 
-        @Override
-        public void operationComplete(ChannelFuture future) throws Exception {
-            Throwable cause = future.cause();
-            if (cause != null) {
-                pipeline().fireExceptionCaught(cause);
-                close();
+        /**
+         * Returns the flow-control size for DATA frames, and 0 for all other frames.
+         */
+        private static final class FlowControlledFrameSizeEstimator implements MessageSizeEstimator {
+
+            static final FlowControlledFrameSizeEstimator INSTANCE = new FlowControlledFrameSizeEstimator();
+
+            private static final class EstimatorHandle implements MessageSizeEstimator.Handle {
+
+                static final EstimatorHandle INSTANCE = new EstimatorHandle();
+
+                @Override
+                public int size(Object msg) {
+                    if (msg instanceof Http2DataFrame) {
+                        return dataFrameFlowControlBytes((Http2DataFrame) msg);
+                    }
+                    return 0;
+                }
+            }
+
+            @Override
+            public Handle newHandle() {
+                return EstimatorHandle.INSTANCE;
             }
         }
-    }
 
-    /**
-     * Wraps the first {@link Http2HeadersFrame} of local/outbound stream. This allows us to get to the child channel
-     * when receiving the {@link Http2StreamActiveEvent} from the frame codec. See {@link #onStreamActive}.
-     */
-    private static final class ChannelCarryingHeadersFrame implements Http2HeadersFrame {
+        /**
+         * {@link ChannelConfig} so that the high and low writebuffer watermarks can reflect the outbound flow control
+         * window, without having to create a new {@link WriteBufferWaterMark} object whenever the flow control window
+         * changes.
+         */
+        private final class Http2StreamChannelConfig extends DefaultChannelConfig {
 
-        private final Http2HeadersFrame frame;
-        private final Http2StreamChannel childChannel;
+            // TODO(buchgr): Overwrite the RecvByteBufAllocator. We only need it to implement max messages per read.
+            Http2StreamChannelConfig(Channel channel) {
+                super(channel);
+            }
 
-        ChannelCarryingHeadersFrame(Http2HeadersFrame frame, Http2StreamChannel childChannel) {
-            this.frame = frame;
-            this.childChannel = childChannel;
-        }
+            /**
+             * @deprecated  Use {@link #getWriteBufferWaterMark()} instead.
+             */
+            @Override
+            @Deprecated
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)